core/msg.c

```diff
@@ -214,10 +214,19 @@ int msg_send_receive(msg_t *m, msg_t *reply, kernel_pid_t target_pid)
     unsigned state = irq_disable();
     thread_t *me = (thread_t*) sched_threads[sched_active_pid];
     sched_set_status(me, STATUS_REPLY_BLOCKED);
-    me->wait_data = (void*) reply;
+    uint8_t target_status = ((thread_t*) sched_threads[target_pid])->status;
 
+    if (target_status == STATUS_RECEIVE_BLOCKED) {
```

Comment:
I think this check might be actually harmful. Since the response is copied later anyway from `me->wait_data`, it should be safe to always keep the assignment in the line below.